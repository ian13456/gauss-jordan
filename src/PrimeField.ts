import { Field } from './Field';
import Utils from './utils';

export class PrimeField extends Field<number> {
  modulus: number;

  constructor(mod: number) {
    super();

    this.modulus = mod;
  }

  zero = () => 0;

  one = () => 1;

  equals = (x: number, y: number) => this._check(x) === this._check(y);

  negate = (x: number) => Utils.mod(-this._check(x), this.modulus);

  add = (x: number, y: number) => Utils.mod(this._check(x) + this._check(y), this.modulus);

  multiply = (x: number, y: number) => Utils.mod(this._check(x) * this._check(y), this.modulus);

  reciprocal = (w: number) => {
    // Extended Euclidean GCD algorithm
    let x = this.modulus;
    let y = this._check(w);
    if (y === 0) throw new Error('Division by zero');

    let a = 0;
    let b = 1;

    while (y !== 0) {
      const q = Math.floor(x / y);
      const r = Utils.mod(x, y);
      x = y;
      y = r;
      const temp = a;
      a = b;
      b = temp - q * b;
    }

    if (x !== 1) throw new Error('Field modulus is not prime');
    return Utils.mod(a, this.modulus);
  };

  _check = (x: number) => {
    if (x !== Math.floor(x)) throw new Error('Number not integer: ' + x);
    if (!(0 <= x && x < this.modulus)) throw new Error('Not an element of this field: ' + x);
    return x;
  };
}
