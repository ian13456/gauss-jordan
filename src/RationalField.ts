import { Field } from './Field';

export class RationalField extends Field<number> {
  static FIELD = new RationalField();

  zero = () => 0;

  one = () => 1;

  equals = (x: number, y: number) => x === y;

  negate = (x: number) => -x;

  add = (x: number, y: number) => x + y;

  reciprocal = (x: number) => 1 / x;

  multiply = (x: number, y: number) => x * y;
}
