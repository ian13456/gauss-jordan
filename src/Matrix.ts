import { Field } from './Field';
import { RationalField } from './RationalField';

export class Matrix<T> {
  f: Field<T>;
  values: T[][];

  constructor(rows: number, cols: number, field: Field<T>) {
    if (rows <= 0 || cols <= 0) throw new Error('Invalid number of rows or columns');

    this.f = field;

    this.values = [];
    for (let i = 0; i < rows; i++) {
      this.values.push(new Array(cols));
    }
  }

  static fromRationalArray(array: number[][]) {
    const f = new RationalField();
    const m = new Matrix<number>(array.length, array[0].length, f);

    for (let i = 0; i < array.length; i++)
      for (let j = 0; j < array[i].length; j++) {
        m.set(i, j, array[i][j]);
      }

    return m;
  }

  rowCount = () => this.values.length;

  columnCount = () => this.values[0].length;

  get = (row: number, col: number) => {
    if (!(0 <= row && row < this.values.length && 0 <= col && col < this.values[row].length))
      throw new Error('Row or column index out of bounds');
    return this.values[row][col];
  };

  set = (row: number, col: number, val: T) => {
    if (!(0 <= row && row < this.values.length && 0 <= col && col < this.values[row].length))
      throw new Error('Row or column index out of bounds');
    this.values[row][col] = val;
  };

  swapRows = (row0: number, row1: number) => {
    if (!(0 <= row0 && row0 < this.values.length && 0 <= row1 && row1 < this.values.length))
      throw new Error('Row index out of bounds');

    const temp = this.values[row0];
    this.values[row0] = this.values[row1];
    this.values[row1] = temp;
  };

  multiplyRow = (row: number, factor: T) => {
    if (!(0 <= row && row < this.values.length)) throw new Error('Row index out of bounds');

    this.values[row] = this.values[row].map((n) => this.f.multiply(n, factor));
  };

  addRows = (srcRow: number, destRow: number, factor: T) => {
    if (!(0 <= srcRow && srcRow < this.values.length && 0 <= destRow && destRow < this.values.length))
      throw new Error('Row index out of bounds');

    this.values[destRow] = this.values[destRow].map((n, i) => {
      const srcNum = this.values[srcRow][i];
      return this.f.add(n, this.f.multiply(srcNum, factor));
    });
  };

  multiply = (other: Matrix<T>) => {
    const rows = this.rowCount();
    const cols = other.columnCount();
    const cells = this.columnCount();
    const result = new Matrix(rows, cols, this.f);

    for (let i = 0; i < rows; i++) {
      for (let j = 0; j < cols; j++) {
        let sum = this.f.zero();
        for (let k = 0; k < cells; k++) sum = this.f.add(this.f.multiply(this.get(i, k), other.get(k, j)), sum);

        result.set(i, j, sum);
      }
    }

    return result;
  };

  // ADVANCED MATRIX OPERATIONS
  reducedRowEchelonForm = () => {
    /*
    Converts this matrix to reduced row echelon form (RREF) using Gauss-Jordan elimination.
		All elements of this matrix should be non-None when performing this operation.
		Always succeeds, as long as the field follows the mathematical rules and does not raise an exception.
		The time complexity of this operation is O(rows * cols * min(rows, cols)).
    */
    const rows = this.rowCount();
    const cols = this.columnCount();

    // Compute row echelon form (REF)
    let numPivots = 0;
    for (let j = 0; j < cols; j++) {
      if (numPivots >= rows) break;

      let pivotRow = numPivots;
      while (pivotRow < rows && this.f.equals(this.get(pivotRow, j), this.f.zero())) {
        pivotRow++;
      }

      if (pivotRow === rows) continue; // Cannot eliminate on this column

      this.swapRows(numPivots, pivotRow);
      pivotRow = numPivots;
      numPivots += 1;

      // Simplify the pivot row
      this.multiplyRow(pivotRow, this.f.reciprocal(this.get(pivotRow, j)));

      // Eliminate rows below
      for (let i = pivotRow + 1; i < rows; i++) {
        this.addRows(pivotRow, i, this.f.negate(this.get(i, j)));
      }
    }

    // Compute reduced row echelon form (RREF)
    for (let i = numPivots - 1; i >= 0; i--) {
      // Find pivot
      let pivotCol = 0;
      while (pivotCol < cols && this.f.equals(this.get(i, pivotCol), this.f.zero())) {
        pivotCol++;
      }
      if (pivotCol === cols) continue; // Skip this all-zero row

      // Eliminate rows above
      for (let j = 0; j < i; j++) {
        this.addRows(i, j, this.f.negate(this.get(j, pivotCol)));
      }
    }

    return this.values;
  };
}
