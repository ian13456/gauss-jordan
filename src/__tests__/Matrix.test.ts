import { PrimeField } from '../PrimeField';
import { RationalField } from '../RationalField';
import { Matrix } from '../Matrix';

describe('Matrix', () => {
  it('should create from rational array', () => {
    const input = [
      [3, 1, 4, 1],
      [5, 2, 6, 5],
      [0, 5, 2, 1],
    ];

    const m = Matrix.fromRationalArray(input);

    expect(m).toMatchSnapshot();
  });
});

describe('PrimeField', () => {
  it('should rref properly', () => {
    const modulus = 7;

    const input = [
      [3, 1, 4, 1],
      [5, 2, 6, 5],
      [0, 5, 2, 1],
    ];

    const p = new PrimeField(modulus);
    const m = new Matrix<number>(input.length, input[0].length, p);

    for (let i = 0; i < input.length; i++)
      for (let j = 0; j < input[i].length; j++) {
        m.set(i, j, input[i][j]);
      }

    m.reducedRowEchelonForm();

    expect(m.values).toMatchSnapshot();
  });
});

describe('RationalField', () => {
  it('should rref properly', () => {
    const input = [
      [2, 5, 3, 7],
      [1, 0, 1, 1],
      [-4, 2, -9, 6],
    ];

    const r = new RationalField();
    const m = new Matrix<number>(input.length, input[0].length, r);

    for (let i = 0; i < input.length; i++)
      for (let j = 0; j < input[i].length; j++) {
        m.set(i, j, input[i][j]);
      }

    m.reducedRowEchelonForm();

    expect(m.values).toMatchSnapshot();
  });
});
