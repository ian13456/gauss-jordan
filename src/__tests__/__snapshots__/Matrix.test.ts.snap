// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix should create from rational array 1`] = `
Matrix {
  "addRows": [Function],
  "columnCount": [Function],
  "f": RationalField {
    "add": [Function],
    "equals": [Function],
    "multiply": [Function],
    "negate": [Function],
    "one": [Function],
    "reciprocal": [Function],
    "zero": [Function],
  },
  "get": [Function],
  "multiply": [Function],
  "multiplyRow": [Function],
  "reducedRowEchelonForm": [Function],
  "rowCount": [Function],
  "set": [Function],
  "swapRows": [Function],
  "values": Array [
    Array [
      3,
      1,
      4,
      1,
    ],
    Array [
      5,
      2,
      6,
      5,
    ],
    Array [
      0,
      5,
      2,
      1,
    ],
  ],
}
`;

exports[`PrimeField should rref properly 1`] = `
Array [
  Array [
    1,
    0,
    0,
    4,
  ],
  Array [
    0,
    1,
    0,
    3,
  ],
  Array [
    0,
    0,
    1,
    0,
  ],
]
`;

exports[`RationalField should rref properly 1`] = `
Array [
  Array [
    1,
    0,
    0,
    2.481481481481482,
  ],
  Array [
    0,
    1,
    0,
    1.2962962962962963,
  ],
  Array [
    -0,
    -0,
    1,
    -1.4814814814814814,
  ],
]
`;
